New generic tableau operator: TableauOperator

mariotaddeucci opened the pull request:
Added an generic tableau operator to execute multiple methods on python client.

mariotaddeucci committed:
Added generic tableau operator

Added generic tableau operator to execute multiple methods on python client

mariotaddeucci committed:
Added generic tableau operator

Added generic tableau operator to execute multiple methods on python client

eladkal commented on airflow/providers/tableau/operators/tableau.py:
```diff
@@ -0,0 +1,124 @@
+    :param site_id: The id of the site where the workbook belongs to.
+    :type site_id: Optional[str]
+    :param blocking: By default the extract refresh will be blocking means it will wait until it has finished.
+    :type blocking: bool
```
There is no `blocking` parameter in the init. Did you mean `blocking_refresh` ?

eladkal commented on airflow/providers/tableau/operators/tableau_refresh_workbook.py:
```diff
@@ -67,18 +68,16 @@ def execute(self, context: dict) -> str:
+                tableau_conn_id=self.tableau_conn_id,
+                task_id='refresh_workbook',
+                dag=None,
+            ).execute(context={})
```
do we actually actually need a `TableauRefreshWorkbookOperator`?
Your code suggestion means that `TableauRefreshWorkbookOperator` is actually equal to just doing:
```
TableauOperator(resource='workbooks',
                method='refresh',
                resource_id='my_workbook_id',
                site_id='site_id',
                tableau_conn_id='tableau_default',
                task_id='refresh_workbook')
```

So maybe a better way is to deprecate `TableauRefreshWorkbookOperator` ?

mariotaddeucci replied:
The main difference is `TableauRefreshWorkbookOperator` uses a find by name and the new uses by id.
Now I add an option to choose what find_by will use, so now the `TableauRefreshWorkbookOperator` is not necessary.
Thanks for the alert

eladkal reviewed (changes requested):
left comments in the code.
Note that also requires tests for `TableauOperator`

mariotaddeucci committed:
Added refresh workbooks and datasources tests

mariotaddeucci committed:
Added refresh workbooks and datasources tests

mariotaddeucci committed:
Added TableauOperator example dag

mariotaddeucci committed:
Added TableauOperator example dag

mariotaddeucci committed:
Deprecated TableauRefreshWorkbookOperato

- Fixed documentation comments on TableauOperator execute
- Added generic find by column
- Deprecated TableauRefreshWorkbookOperator

mariotaddeucci committed:
Deprecated TableauRefreshWorkbookOperato

- Fixed documentation comments on TableauOperator execute
- Added generic find by column
- Deprecated TableauRefreshWorkbookOperator

eladkal commented on airflow/providers/tableau/operators/tableau_refresh_workbook.py:
```diff
@@ -65,31 +76,16 @@ def execute(self, context: dict) -> str:
+                tableau_conn_id=self.tableau_conn_id,
+                task_id='refresh_workbook',
+                dag=None,
+            ).execute(context={})
```
Is it backward compatible? The `TableauRefreshWorkbookOperator` had `blocking`, the TableauOperator has `blocking_refresh`. This means that it will be broken for some users.

eladkal commented on airflow/providers/tableau/operators/tableau.py:
```diff
@@ -0,0 +1,144 @@
+            response = method(resource_id)
+
+        if self.method == 'download':
+            return response
```
If I understand it right this means that you push to xcom the downloaded data source. I'm not tableau expert but isn't it may be massive amount of data? Should we apply limitation to it like `MAX_XCOM_SIZE` ? example in `GCSToLocalFilesystemOperator`: 

https://github.com/apache/airflow/blob/866a601b76e219b3c043e1dbbc8fb22300866351/airflow/providers/google/cloud/transfers/gcs_to_local.py#L134-L140

(Also another thought maybe we don't need to return anything when `method == 'download' `?)
WDYT?

mariotaddeucci replied:
Whe using the download method, it push the file URI to xcoms. Usually is `/temp/the-tableau-file`.
The URI is not a huge string and doing that it enables you to use S3Operstor to backup the downloaded file

eladkal replied:
OK got it. But how is it helpful to a downstream task? Tasks may run on diffrent workers thus having diffrent disks.

mariotaddeucci replied:
Yeah, this is a problem. I'll remove this method for now

eladkal commented on airflow/providers/tableau/provider.yaml:
```diff
@@ -37,6 +37,7 @@ integrations:
 operators:
   - integration-name: Tableau
     python-modules:
+      - airflow.providers.tableau.operators.tableau
```
Would be nice also to add docs about the operator usage.
This needs to be added to:
https://github.com/apache/airflow/tree/main/docs/apache-airflow-providers-tableau

Examples from other operators:
https://github.com/apache/airflow/blob/main/docs/apache-airflow-providers-amazon/operators/emr.rst
https://github.com/apache/airflow/blob/main/docs/apache-airflow-providers-mysql/operators.rst
https://github.com/apache/airflow/blob/main/docs/apache-airflow-providers-asana/operators/asana.rst

mariotaddeucci replied:
Done

eladkal commented on airflow/providers/tableau/operators/tableau.py:
```diff
@@ -0,0 +1,144 @@
+        resource: str,
+        method: str,
+        resource_id: str,
+        resource_id_by: str = 'id',
```
What it means when you provide both `resource_id` & `resource_id_by` (Like shown in the example dag you added)? From what I see you need only 1 of them (as if user provided `resource_id_by` then you search for the id and then execute the method ). It can be confusing when both are provided isn't it?

mariotaddeucci replied:
The tableu client uses the tableu an UUID for operations. To be compatible with `TableauRefreshWorkbookOperator` that uses the workbooks name to get this UUID then to execute the tasks, the resource_id_by says what resource field should be use to match with resource_id to return the UUID necessary to run the tasks.
Agree, it's confusing. Should I change it to `find` and `find_by`?
Wich `find_by` will recive the resource field name to match and `find` the value to match.

eladkal replied:
The confusing part I was referring to isn't the parameter names.
When you provide both `resource_id` & `resource_id_by` the value in `resource_id_by` will be ignored because of the condition in `_get_resource_id_by` function. I suggest to make both `resource_id` & `resource_id_by` optional parameters and make sure the user pass exactly one of them.
We can make sure the two parameters are mutually exclusive.  WDYT?

mariotaddeucci replied:
Both are necessary.  
The method `_get_resource_id_by` always is executed.
When you pass `resource_id='foo'`  and `resource='workbooks'` It will search an workbook with ID foo. 
When you pass `resource_id='foo'`, `resource_id_by='name'`  and `resource='workbooks'` It will search an workbook with name foo.
When you pass `resource_id_by='name'`, `resource='workbooks'` and `resource_id=None` It won't work because the search happen using `getattr(resource, self.resource_id_by) == self.resource_id`

That's why I ask about change to find and find_by. 

```
    def _get_resource_id_by(self, tableau_hook: TableauHook) -> str:

        if self.resource_id_by == 'id':
            return self.resource_id

        for resource in tableau_hook.get_all(resource_name=self.resource):
            if getattr(resource, self.resource_id_by) == self.resource_id:
                resource_id = resource.id
                self.log.info('Found matching with id %s', resource_id)
                return resource_id
```

mariotaddeucci replied:
Guess now it's clearer.
**resource**: The name of the resource to use.
**method**: The name of the resource's method to execute.
**find**: The reference of resource which will receive the action.
**match_with**: The resource field name to be matched with **find** parameter.

mariotaddeucci committed:
Added doc and fixed operator parameters

mariotaddeucci committed:
Added doc and fixed operator parameters

mariotaddeucci committed:
Update operators.rst

mariotaddeucci committed:
Update operators.rst

mariotaddeucci committed:
Update operators.rst

mariotaddeucci committed:
Update operators.rst

mariotaddeucci committed:
Update operators.rst

mariotaddeucci committed:
Update operators.rst

mariotaddeucci committed:
Fixed block refresh on TableauRefreshWorkbookOperator

mariotaddeucci committed:
Fixed block refresh on TableauRefreshWorkbookOperator

mariotaddeucci committed:
Resolve conflicts

mariotaddeucci committed:
Merge branch 'main' of https://github.com/mariotaddeucci/airflow into main

mariotaddeucci committed:
Removed unused tableau_hook on TableauRefreshWorkbookOperator

mariotaddeucci committed:
Merge branch 'apache:main' into main

mariotaddeucci committed:
Fixed example dag

mariotaddeucci committed:
Merge branch 'main' into main

potiuk reviewed (approved):
LGTM :). @eladkal ?

eladkal commented on airflow/providers/tableau/operators/tableau.py:
```diff
@@ -0,0 +1,140 @@
+                    return_code = int(tableau_hook.server.jobs.get_by_id(job_id).finish_code)
+                    finish_code = TableauJobFinishCode(return_code)
+                    if finish_code in negative_codes:
+                        raise TableauJobFailedException(f'The Tableau Refresh {self.resource} Job failed!')
```
The hook has `wait_for_state` function which was recently added. Why not using it?

mariotaddeucci replied:
@eladkal nice. I changed to use that  now.

mariotaddeucci committed:
Removed duplicated code

eladkal commented on airflow/providers/tableau/operators/tableau.py:
```diff
@@ -0,0 +1,139 @@
+            if self.blocking_refresh:
+                if not tableau_hook.wait_for_state(
+                    job_id=job_id,
+                    check_interval=self.check_interval,
```
`self.check_interval `isn't defined.
You need to add it to the constactor (don't forget also to add it in the class description.

mariotaddeucci replied:
Fixed

mariotaddeucci committed:
Fixed check_interval param

mariotaddeucci committed:
Moded tests from tableau_workbook_operator to tableau_operator

mariotaddeucci committed:
Merge branch 'apache:main' into main

mariotaddeucci committed:
Fixed unit tests

mariotaddeucci committed:
Fixed flake8

mariotaddeucci committed:
Lint fix

mariotaddeucci committed:
Isort fixeds

mariotaddeucci committed:
Fixed static checks

mariotaddeucci committed:
Update operators.rst

eladkal commented on docs/apache-airflow-providers-tableau/operators.rst:
```diff
@@ -0,0 +1,101 @@
+| **site_id**: The id of the site where the workbook belongs to. **str** - Default: **None**
+| **blocking_refresh**: By default the extract refresh will be blocking means it will wait until it has finished. **bool** - Default: **True**
+| **check_interval**: time in seconds that the job should wait in between each instance state checks until operation is completed. **float** - Default: **20**
+| **tableau_conn_id**: The credentials to authenticate to the Tableau Server. **str** - Default: **tableau_default**  
```
```suggestion
| **tableau_conn_id**: The credentials to authenticate to the Tableau Server. **str** - Default: **tableau_default**
```
Remove white space to make static checks pass

mariotaddeucci committed:
Fixed static checks

mariotaddeucci committed:
Merge branch 'apache:main' into main

mariotaddeucci committed:
Fix Lint and doc example reference

mariotaddeucci committed:
Fix doc toctree

mariotaddeucci committed:
Spelling fixes

mariotaddeucci committed:
Fix spell

mariotaddeucci committed:
Merge branch 'apache:main' into main

eladkal commented on docs/apache-airflow-providers-tableau/operators.rst:
```diff
@@ -0,0 +1,101 @@
+
+
+
+.. list-table:: Available methods by resource
```
Would be nice if you can center the columns values
Currently it looks like:
![Screen Shot 2021-08-08 at 20 14 31](https://user-images.githubusercontent.com/45845474/128640017-e5014e48-555d-42e5-95b5-9c6f07e5f352.png)

mariotaddeucci replied:
I didn’t find a way to center it.
I changed the table format to be clearer 
![image](https://user-images.githubusercontent.com/8824036/128642609-c36252ff-0673-4698-898d-a37abb7e409c.png)

eladkal reviewed (approved):
LGTM (left one comment on but it's not that important)

mariotaddeucci committed:
Improve doc methods table

mariotaddeucci committed:
Merge branch 'apache:main' into main

mariotaddeucci committed:
Merge branch 'apache:main' into main
